Jetpack Components: Add Toggle

nateweller opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Adds a `<ToggleControl />` component to `@automattic/jetpack-components`.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

TTzDk3dzct551hYWRXPFa3-fi-2394%3A28057

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Check out this branch
* Run `jetpack install -r`
* Run `cd projects/js-packages/storybook`
* Run `pnpm run storybook:dev`
* Review the "Toggle Control" component stories and compare them to the Figma designs.

## Screenshots

<img width="1132" alt="Screenshot 2023-04-18 at 4 42 41 PM" src="https://user-images.githubusercontent.com/10933065/232920787-055c1f66-8a25-4878-9c29-6d6cd2cdcfc1.png">

nateweller committed:
[not verified] Add a Toggle component to the jetpack-components package

nateweller committed:
[not verified] Changelog

nateweller committed:
[not verified] Fixup project versions

nateweller committed:
[not verified] Update README

renatoagds commented:
@nateweller 

This one implements the same `Toggle` we have in [core](https://automattic.github.io/jetpack-storybook/?path=/story/gutenberg_components-togglecontrol--default&globals=measureEnabled:false;outline:false) with colors improvements. I suggest to use it as the base in this one instead implementing everything. This is the current approach that we have been using in `Jetpack Components.`

This one already has been discussed to be added on top of `ToggleControl` from the core:

p1664542252556529-slack-C02M27K4741

And suggest to ping us in `#jetpack-rna` when opening some PR in `Jetpack Components` cc/ @Automattic/jetpack-crew 

Example how we do it here:

p1678881332031789-slack-C02M27K4741

anomiex commented on projects/js-packages/components/components/toggle/stories/index.jsx:
This will need to be renamed for Storybook 7 (probably to `.../index.stories.jsx`). See 51eaa5009b6b5b2f82806cf609c9a65c1aa684d1 for more info.

anomiex commented on projects/js-packages/components/components/toggle/stories/index.jsx:
```diff
@@ -0,0 +1,60 @@
+	},
+};
+
+export const Default = () => {
```
The "Show code" option for this story won't show the correct code with Storybook 7, it seems that an `args` parameter to the function is now required for that to work. See 51eaa5009b6b5b2f82806cf609c9a65c1aa684d1 for more info.

Same goes for other stories below.

nateweller commented:
> This one implements the same Toggle we have in [core](https://automattic.github.io/jetpack-storybook/?path=/story/gutenberg_components-togglecontrol--default&globals=measureEnabled:false;outline:false) with colors improvements. I suggest to use it as the base in this one instead implementing everything. This is the current approach that we have been using in Jetpack Components.

@renatoagds Thank you for pointing this out - my mistake. I can adjust this PR so that we build on top of the FormToggle from `@wordpress/components`. :+1:

I've joined the Slack channel as well!

> This will need to be renamed for Storybook 7

@anomiex Cool, I will update accordingly! We've just started using Storybook 7 for WPScan as well, and it has been pretty nice so far.

nateweller committed:
[not verified] Use @wordpress/components ToggleControl as a base for the Jetpack component

nateweller commented:
Updated - now uses the [same approach as Social](https://github.com/Automattic/jetpack/tree/trunk/projects/plugins/social/src/js/components/toggle-section) to apply RNA styles to the component from core. Stories compatible with Storybook 7 as well :+1:

nateweller committed:
[not verified] Update styles to override core component by specificity

renatoagds commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php:
```diff
@@ -66,6 +66,32 @@ function get_checklist_definitions() {
+function can_update_design_selected_task() {
+	$site_intent = get_option( 'site_intent' );
+	return $site_intent === 'free' || $site_intent === 'build' || $site_intent === 'write';
+}
```
This one looks like a bad merge, since isn't related with the PR, right?

nateweller replied:
Whoops, must have forgotten to rebase after switching the base branch. Fixed that! :+1:

renatoagds commented on projects/js-packages/components/components/toggle-control/index.tsx:
```diff
@@ -0,0 +1,45 @@
+import { ToggleControl as CalypsoToggleControl } from '@wordpress/components';
```
```suggestion
import { ToggleControl as WPToggleControl } from '@wordpress/components';
```

It isn't Calypso but WP.org

nateweller committed:
[not verified] Add toggling state and rename core component

nateweller commented:
Addressed comments from @renatoagds and also added a `toggling` prop to eagerly switch the toggle value as soon as it is clicked, and prevent additional clicks while the toggle is in progress, without disabling the input and losing focus.

See the "Slow" story for an example. :+1:

dhasilva reviewed (approved):
LGTM. Thanks for adding this component!
